kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java

```diff
@@ -230,48 +232,70 @@ public void handleInitProducerId(String transactionalId,
-                int coordinatorEpoch = epochAndTxnMeta.getData().get().getCoordinatorEpoch();
-                TransactionMetadata txnMetadata = epochAndTxnMeta.getData().get().getTransactionMetadata();
-
-                txnMetadata.inLock(() -> {
-                    CompletableFuture<ErrorsAndData<EpochAndTxnTransitMetadata>> prepareInitProducerIdResult =
-                            prepareInitProducerIdTransit(
-                                transactionalId,
-                                transactionTimeoutMs,
-                                coordinatorEpoch,
-                                txnMetadata,
-                                expectedProducerIdAndEpoch);
-
-                    prepareInitProducerIdResult.whenComplete((errorsAndData, prepareThrowable) -> {
-                        completeInitProducer(
-                                transactionalId, coordinatorEpoch, errorsAndData, prepareThrowable, responseCallback);
+            epochAndTxnMetaFuture.thenAccept(epochAndTxnMeta -> {
+                if (epochAndTxnMeta.hasErrors()) {
+                    responseCallback.accept(initTransactionError(epochAndTxnMeta.getErrors()));
+                    return;
+                }
+                Optional<CoordinatorEpochAndTxnMetadata> data = epochAndTxnMeta.getData();
+                if (data.isPresent()) {
+                    int coordinatorEpoch = data.get().getCoordinatorEpoch();
+                    TransactionMetadata txnMetadata = epochAndTxnMeta.getData().get().getTransactionMetadata();
+
+                    txnMetadata.inLock(() -> {
+                        CompletableFuture<ErrorsAndData<EpochAndTxnTransitMetadata>> prepareInitProducerIdResult =
+                                prepareInitProducerIdTransit(
+                                        transactionalId,
+                                        transactionTimeoutMs,
+                                        coordinatorEpoch,
+                                        txnMetadata,
+                                        expectedProducerIdAndEpoch);
+
+                        prepareInitProducerIdResult.whenComplete((errorsAndData, prepareThrowable) -> {
+                            completeInitProducer(
+                                    transactionalId,
+                                    coordinatorEpoch,
+                                    errorsAndData,
+                                    prepareThrowable,
+                                    responseCallback);
+                        });
+                        return null;
                     });
-                    return null;
-                });
+                } else {
+                    responseCallback.accept(initTransactionError(Errors.UNKNOWN_SERVER_ERROR));
+                }
+            }).exceptionally(ex -> {
+                responseCallback.accept(initTransactionError(Errors.forException(ex.getCause())));
```

Comment:
Just a trivial comment, maybe we could add an error log here.

Reply:
You'r right. Addressed.